### PR TITLE
Add input plugin for DC/OS

### PR DESCRIPTION
**Pull request by danielnelson**
### Required for all PRs:

- [x] Signed [CLA](https://influxdata.com/community/cla/).
- [x] Associated README.md updated.
- [x] Has appropriate unit tests.

### CODE REVIEW

File: plugins/inputs/dcos/client.go [hunk 1]
```diff
@@ -0,0 +1,340 @@
...
+	PrivateKey *rsa.PrivateKey
+	TokenFile  string
+}
+
+type APIError struct {
+	StatusCode  int
+	Title       string
+	Description string
+}
+type Login struct {
```
Comment (goller) [inline plugins/inputs/dcos/client.go]: newline here

File: plugins/inputs/dcos/client.go [hunk 2]
```diff
@@ -0,0 +1,340 @@
...
+	Token() string
+	EnsureAuth(ctx context.Context) error
+	GetSummary(ctx context.Context) (*Summary, error)
+	GetContainers(ctx context.Context, node string) ([]string, error)
+	GetNodeMetrics(ctx context.Context, node string) (*Metrics, error)
+	GetContainerMetrics(ctx context.Context, node, container string) (*Metrics, error)
+	GetAppMetrics(ctx context.Context, node, container string) (*Metrics, error)
+}
+
+type Credentials struct {
```
Comment (goller) [inline plugins/inputs/dcos/client.go]: Would be nice to have comments on structs

File: plugins/inputs/dcos/client.go [hunk 3]
```diff
@@ -0,0 +1,340 @@
...
+
+type Summary struct {
+	Cluster string
+	Slaves  []Slave
+}
+
+type DataPoint struct {
+	Name  string
+	Tags  map[string]string
+	Unit  string
```
Comment (goller) [inline plugins/inputs/dcos/client.go]: perhaps make optional fields pointers?

File: plugins/inputs/dcos/client.go [hunk 4]
```diff
@@ -0,0 +1,340 @@
...
+	err := c.doGet(ctx, c.url(path), &containers)
+	if err != nil {
+		return nil, err
+	}
+
+	return containers, nil
+}
+
+func (c *client) GetNodeMetrics(ctx context.Context, node string) (*Metrics, error) {
+	metrics := &Metrics{}
```
Comment (goller) [inline plugins/inputs/dcos/client.go]: Perhaps add a getMetrics helper that takes a path.

File: plugins/inputs/dcos/client.go [hunk 5]
```diff
@@ -0,0 +1,340 @@
...
+
+	resp, err := c.httpClient.Do(req.WithContext(ctx))
+	if err != nil {
+		<-c.semaphore
+		return err
+	}
+	defer resp.Body.Close()
+
+	// Clear invalid token if unauthorized
+	if resp.StatusCode == 401 {
```
Comment (goller) [inline plugins/inputs/dcos/client.go]: use the http status codes.

File: plugins/inputs/dcos/client.go [hunk 6]
```diff
@@ -0,0 +1,340 @@
...
+	if resp.StatusCode == 204 {
+		<-c.semaphore
+		return nil
+	}
+
+	err = json.NewDecoder(resp.Body).Decode(v)
+	<-c.semaphore
+	return err
+}
+func (c *client) url(path string) string {
```
Comment (goller) [inline plugins/inputs/dcos/client.go]: add newline

File: plugins/inputs/dcos/client.go [hunk 7]
```diff
@@ -0,0 +1,340 @@
...
+
+func (c *client) login(ctx context.Context) (*authToken, error) {
+	token, err := c.createLoginToken()
+	if err != nil {
+		return nil, err
+	}
+
+	exp := time.Now().Add(loginDuration)
+
+	body := map[string]interface{}{
```
Comment (goller) [inline plugins/inputs/dcos/client.go]: could be an anonymous struct.

File: plugins/inputs/dcos/client.go [hunk 8]
```diff
@@ -0,0 +1,340 @@
...
+	defer resp.Body.Close()
+
+	login := Login{}
+	dec := json.NewDecoder(resp.Body)
+	err = dec.Decode(&login)
+	if err != nil {
+		return nil, err
+	}
+
+	if resp.StatusCode != 200 || login.Token == "" {
```
Comment (goller) [inline plugins/inputs/dcos/client.go]: Decode body after checking status to decode into different structs.

File: plugins/inputs/dcos/client.go [hunk 9]
```diff
@@ -0,0 +1,340 @@
...
+		return nil, &APIError{resp.StatusCode, login.Title, login.Description}
```
Comment (goller) [inline plugins/inputs/dcos/client.go]: Add struct field names.

File: plugins/inputs/dcos/client.go [hunk 10]
```diff
@@ -0,0 +1,340 @@
...
+}
+
+func (c *client) createLoginToken() (string, error) {
+	token := jwt.NewWithClaims(jwt.SigningMethodRS256, claims{
+		Uid: c.credentials.Username,
+		StandardClaims: jwt.StandardClaims{
+			ExpiresAt: 0,
+		},
+	})
+	ss, err := token.SignedString(c.credentials.PrivateKey)
```
Comment (goller) [inline plugins/inputs/dcos/client.go]: Just return here.

File: plugins/inputs/dcos/dcos.go [hunk 11]
```diff
@@ -0,0 +1,410 @@
...
+  #   path = ["/var/lib/mesos/slave/slaves/*"]
+`
+
+func (d *DCOS) SampleConfig() string {
+	return sampleConfig
+}
+
+func (d *DCOS) Gather(acc telegraf.Accumulator) error {
+	if !d.initialized {
+		err := d.createFilters()
```
Comment (goller) [inline plugins/inputs/dcos/dcos.go]: put initialized in createFilters

File: plugins/inputs/dcos/dcos.go [hunk 12]
```diff
@@ -0,0 +1,410 @@
...
+		}
+
+		if dp.Unit == "bytes" {
+			p.fields[fieldKey] = int64(dp.Value)
+		} else {
+			p.fields[fieldKey] = dp.Value
+		}
+	}
+
+	results := []*point{}
```
Comment (goller) [inline plugins/inputs/dcos/dcos.go]: preallocate the slice.

File: plugins/inputs/dcos/dcos.go [hunk 13]
```diff
@@ -0,0 +1,410 @@
...
+		tags := dp.Tags
+		if tags == nil {
+			tags = make(map[string]string)
+		}
+
+		if dp.Unit == "bytes" && !strings.HasSuffix(fieldKey, "_bytes") {
+			fieldKey = fieldKey + "_bytes"
+		}
+
+		tagset := []string{}
```
Comment (goller) [inline plugins/inputs/dcos/dcos.go]: preallocate here.

File: plugins/inputs/dcos/dcos.go [hunk 14]
```diff
@@ -0,0 +1,410 @@
...
+			}
+		}
+		for k, v := range p.labels {
+			tags[k] = v
+		}
+
+		fields := make(map[string]interface{})
+		for k, v := range p.fields {
+			if strings.HasPrefix(k, "dcos_metrics_module_") {
+				k = strings.TrimPrefix(k, "dcos_metrics_module_")
```
Comment (goller) [inline plugins/inputs/dcos/dcos.go]: Probably move this trim up above.

**Comment by danielnelson** [comment]
I think everything is addressed except the struct comments.

**Pull request merged**